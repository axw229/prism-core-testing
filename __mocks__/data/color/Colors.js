// @flow
import cloneDeep from 'lodash/cloneDeep'
import sample from 'lodash/sample'

import type { ColorSetPayload, Color } from '../../../src/shared/types/Colors'

export const Colors: ColorSetPayload = {
  "Red": [
    [
      {
        "colorNumber": "6561",
        "coordinatingColors": {
          "coord2ColorId": 1926,
          "coord1ColorId": 2250,
          "whiteColorId": 2681
        },
        "description": [
          "Subdued",
          "Muted",
          "Brilliant",
          "Radiant"
        ],
        "id": "2248",
        "isExterior": true,
        "isInterior": true,
        "name": "Teaberry",
        "rgb": 15454683,
        "lrv": 68.551,
        "colorFamilyNames": [
          "Red"
        ],
        "brandKey": "SW",
        "red": 235,
        "green": 209,
        "blue": 219,
        "hue": 0.935897435897436,
        "saturation": 0.39393939393939387,
        "lightness": 0.8705882352941177,
        "hex": "#ebd1db",
        "isDark": false,
        "cssrgb": "rgb(235,209,219)",
        "storeStripLocator": "102-C1",
        "similarColors": [
          "2255",
          "2262",
          "2269",
          "2662",
          "2535",
          "2276",
          "2658",
          "1974",
          "2543",
          "2531"
        ]
      },
      {
        "colorNumber": "6568",
        "coordinatingColors": {
          "coord2ColorId": 11244,
          "coord1ColorId": 2687,
          "whiteColorId": 2681
        },
        "description": [
          "Subdued",
          "Muted",
          "Brilliant",
          "Radiant"
        ],
        "id": "2255",
        "isExterior": true,
        "isInterior": true,
        "name": "Lighthearted Pink",
        "rgb": 15586781,
        "lrv": 70.758,
        "brandedCollectionIds": [
          31797,
          31791
        ],
        "brandedCollectionNames": [
          "Pottery Barn Fall/Winter 2018",
          "Pottery Barn Fall/Winter 2018"
        ],
        "colorFamilyNames": [
          "Red"
        ],
        "brandKey": "SW",
        "red": 237,
        "green": 213,
        "blue": 221,
        "hue": 0.9444444444444445,
        "saturation": 0.3999999999999999,
        "lightness": 0.8823529411764706,
        "hex": "#edd5dd",
        "isDark": false,
        "cssrgb": "rgb(237,213,221)",
        "storeStripLocator": "103-C1",
        "similarColors": [
          "2248",
          "2262",
          "2269",
          "2276",
          "1974",
          "1981",
          "2658",
          "1995",
          "2662",
          "1988"
        ]
      }
    ]
  ]
}

export function getColors (): ColorSetPayload {
  return cloneDeep(Colors)
}

export function getColor (): Color | void {
  const colors = getColors()
  const fam1 = Object.keys(colors)[0]
  if (colors[fam1]) {
    return sample(colors[fam1][0])
  }
}
