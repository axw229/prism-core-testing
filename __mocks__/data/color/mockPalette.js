const mockPalette = {
  'colors': [
    {
      'colorNumber': '6765',
      'coordinatingColors': {
        'coord2ColorId': '11194',
        'coord1ColorId': '2870',
        'whiteColorId': '2162'
      },
      'description': [
        'Fairly Colorful',
        'Moderately Colorful',
        'Pale'
      ],
      'id': '2452',
      'isExterior': true,
      'isInterior': true,
      'name': 'Spa',
      'lrv': 64.476,
      'brandedCollectionNames': [],
      'colorFamilyNames': [
        'Blue'
      ],
      'brandKey': 'SW',
      'red': 167,
      'green': 220,
      'blue': 220,
      'hue': 0.5,
      'saturation': 0.43089430894308944,
      'lightness': 0.7588235294117647,
      'hex': '#a7dcdc',
      'isDark': false,
      'storeStripLocator': '164-C2',
      'similarColors': [
        '2634',
        '2630',
        '11225',
        '2472',
        '2445',
        '2622',
        '2459',
        '1606',
        '11226',
        '2451'
      ],
      'ignore': false,
      'archived': false
    },
    {
      'colorNumber': '6860',
      'coordinatingColors': {
        'coord2ColorId': '11339',
        'coord1ColorId': '11364',
        'whiteColorId': '2681'
      },
      'description': [
        'Soft',
        'Fairly Bright'
      ],
      'id': '2547',
      'isExterior': true,
      'isInterior': true,
      'name': 'Eros Pink',
      'lrv': 18.874,
      'brandedCollectionNames': [
        '2020 Play',
        '2019 Naturalist'
      ],
      'colorFamilyNames': [
        'Red'
      ],
      'brandKey': 'SW',
      'red': 200,
      'green': 79,
      'blue': 104,
      'hue': 0.9655647382920111,
      'saturation': 0.5238095238095238,
      'lightness': 0.5470588235294118,
      'hex': '#c84f68',
      'isDark': false,
      'storeStripLocator': '105-C5',
      'similarColors': [
        '2273',
        '2266',
        '2542',
        '2546',
        '1998',
        '2280',
        '2530',
        '2287',
        '2279',
        '30193'
      ],
      'ignore': false,
      'archived': false
    },
    {
      'colorNumber': '9018',
      'coordinatingColors': {
        'coord2ColorId': '11218',
        'coord1ColorId': '11200',
        'whiteColorId': '2366'
      },
      'description': [
        'Intense',
        'Pale'
      ],
      'id': '11195',
      'isExterior': false,
      'isInterior': true,
      'name': 'Honey Bees',
      'lrv': 70.14,
      'brandedCollectionNames': [
        '2018 Unity'
      ],
      'colorFamilyNames': [
        'Yellow'
      ],
      'brandKey': 'SW',
      'red': 251,
      'green': 214,
      'blue': 130,
      'hue': 0.11570247933884299,
      'saturation': 0.937984496124031,
      'lightness': 0.7470588235294118,
      'hex': '#fbd682',
      'isDark': false,
      'storeStripLocator': '134-C4',
      'similarColors': [
        '2382',
        '11199',
        '11196',
        '2588',
        '2375',
        '11198',
        '11194',
        '11191',
        '3146',
        '2361'
      ],
      'ignore': false,
      'archived': false
    },
    {
      'colorNumber': '9007',
      'coordinatingColors': {
        'coord2ColorId': '1666',
        'coord1ColorId': '2686',
        'whiteColorId': '2331'
      },
      'description': [
        'Rich',
        'Vibrant',
        'Pale'
      ],
      'id': '11184',
      'isExterior': false,
      'isInterior': true,
      'name': 'Melón Meloso',
      'lrv': 55.314,
      'brandedCollectionNames': [],
      'colorFamilyNames': [
        'Orange'
      ],
      'brandKey': 'SW',
      'red': 242,
      'green': 184,
      'blue': 140,
      'hue': 0.0718954248366013,
      'saturation': 0.7968749999999999,
      'lightness': 0.7490196078431373,
      'hex': '#f2b88c',
      'isDark': false,
      'storeStripLocator': '121-C3',
      'similarColors': [
        '2334',
        '2983',
        '2327',
        '11188',
        '2326',
        '2045',
        '2038',
        '3012',
        '2341',
        '2031'
      ],
      'ignore': false,
      'archived': false
    }
  ],
  'activeColor': {
    'colorNumber': '9007',
    'coordinatingColors': {
      'coord2ColorId': '1666',
      'coord1ColorId': '2686',
      'whiteColorId': '2331'
    },
    'description': [
      'Rich',
      'Vibrant',
      'Pale'
    ],
    'id': '11184',
    'isExterior': false,
    'isInterior': true,
    'name': 'Melón Meloso',
    'lrv': 55.314,
    'brandedCollectionNames': [],
    'colorFamilyNames': [
      'Orange'
    ],
    'brandKey': 'SW',
    'red': 242,
    'green': 184,
    'blue': 140,
    'hue': 0.0718954248366013,
    'saturation': 0.7968749999999999,
    'lightness': 0.7490196078431373,
    'hex': '#f2b88c',
    'isDark': false,
    'storeStripLocator': '121-C3',
    'similarColors': [
      '2334',
      '2983',
      '2327',
      '11188',
      '2326',
      '2045',
      '2038',
      '3012',
      '2341',
      '2031'
    ],
    'ignore': false,
    'archived': false
  },
  'toggleCompareColor': false,
  'temporaryActiveColor': null
}

export default mockPalette
