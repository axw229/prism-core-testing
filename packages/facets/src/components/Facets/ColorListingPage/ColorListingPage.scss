.transition-wrapper {
  width: 100%;
  overflow: hidden;
}

.inline-color-details {
  &.enter { transform: translateX(100%); }
  &.enter-active { transform: translateX(0%); }
  &.exit { transform: translateX(0%); }
  &.exit-active { transform: translateX(100%); }
  &.enter-active, &.exit-active { transition: transform 200ms; }
}

.color-wall-wrap {
  &.enter  { opacity: 0; }
  &.enter-active { opacity: 1; }
  &.exit { opacity: 1; }
  &.exit-active { opacity: 0; }
  &.enter-active, &.exit-active { transition: opacity 100ms; }
}