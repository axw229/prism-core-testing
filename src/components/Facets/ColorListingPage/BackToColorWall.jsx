// @flow
import React, { PureComponent } from 'react'
import { Link } from 'react-router-dom'
import { FormattedMessage } from 'react-intl'
import { connect } from 'react-redux'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

import { generateColorWallPageUrl, fullColorName } from '../../../shared/helpers/ColorUtils'
import type { Color } from '../../../shared/types/Colors'

type StateProps = {
  color?: Color,
  family?: string,
  section?: string
}

type OwnProps = {}

type Props = StateProps & OwnProps

class BackToColorWall extends PureComponent<Props> {
  render () {
    const { section, family, color } = this.props

    // this will degrade gracefully if section and/or family and or color do(es) not exist in redux down to the root color wall URL
    const colorWallUrl = color
      ? generateColorWallPageUrl(section, family, color.id, fullColorName(color.brandKey, color.colorNumber, color.name))
      : generateColorWallPageUrl(section, family)

    return (
      <div className='color-wall-mode-btns'>
        <div className='color-wall-mode-btns__col'>
          <div className='color-wall-mode-btns__cell'>
            <FormattedMessage id='BACK_TO_COLOR_WALL'>
              {(msg: string) => (
                <Link to={colorWallUrl} className='color-wall-mode-btns__btn'>
                  <FontAwesomeIcon icon={['fal', 'th-large']} pull='left' fixedWidth />
                  {msg}
                </Link>
              )}
            </FormattedMessage>
          </div>
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => {
  return {
    family: state.colors.family.family,
    section: state.colors.family.section,
    color: state.colors.colorWallActive
  }
}

export default connect(mapStateToProps, null)(BackToColorWall)
