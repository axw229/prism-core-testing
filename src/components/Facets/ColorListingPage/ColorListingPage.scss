// ------------------------------------------------------------------
// BEGIN COLOR WALL <-> COLOR DETAILS TRANSITION

@keyframes slideDetailsIn {
  0%, 20% {
    opacity: 0;
    transform: translate3d(50%, 0, 0);
  }
  80%, 100% {
    opacity: 1;
    transform: translate3d(0%, 0, 0);
  }
}

@keyframes slideWallIn {
  0%, 20% {
    opacity: 0;
    transform: translate3d(-50%, 0, 0);
  }
  80%, 100% {
    opacity: 1;
    transform: translate3d(0%, 0, 0);
  }
}

.color-wall-transitioner {
  overflow: hidden;
  position: relative;

  .cdp-slide {
    &-enter,
    &-enter-active,
    &-enter-done {
      // keep duration significatnly less than the transitionTime that JS is using
      // to reduce jumpy transitions and background flashing in Chrome
      animation-duration: #{getVar('colorWall.transitionTime')}ms;
      animation-fill-mode: both;
      animation-timing-function: cubic-bezier(0.25, 0.46, 0.45, 0.94);
      backface-visibility: hidden;
      background: white;
      position: relative;
      z-index: 100;
      will-change: transform opacity;

      &.cdp-slide-to-wall {
        animation-name: slideWallIn;
      }

      &.cdp-slide-to-details {
        animation-name: slideDetailsIn;
      }
    }

    &-enter-done {
      animation: none;
      transform: unset;
    }

    &-exit,
    &-exit-active,
    &-exit-done {
      left: 0;
      position: absolute;
      top: 0;
      transform: unset;
      width: 100%;
      z-index: 0;
    }

    &-exit-done {
      display: none;
    }
  }
}

// END COLOR WALL <-> COLOR DETAILS TRANSITION
// ------------------------------------------------------------------

// ------------------------------------------------------------------
// BEGIN COLOR WALL <-> COLOR SWATCH ACTIVE TRANSITION

$zoom-out-amt: .9;
$zoom-out-opacity: 0;
$zoom-in-amt: 1;
$zoom-in-opacity: 1;

@keyframes zoomSwatchIn {
  0%, 20% {
    transform: scale($zoom-out-amt);
    opacity: $zoom-out-opacity;
  }
  80%, 100% {
    transform: scale($zoom-in-amt);
    opacity: $zoom-in-opacity;
  }
}

@keyframes zoomSwatchOut {
  0%, 20% {
    transform: scale($zoom-in-amt);
    opacity: $zoom-in-opacity;
  }
  80%, 100% {
    transform: scale($zoom-out-amt);
    opacity: $zoom-out-opacity;
  }
}

.color-wall-zoom-transitioner {
  overflow: hidden;
  position: relative;

  &__zoom-in {
    &--enter,
    &--enter-active,
    &--enter-done {
      // zoomed color wall appearing
      backface-visibility: hidden;
      position: relative;
      z-index: 1;
      will-change: transform;

      animation-name: zoomSwatchIn;
      animation-fill-mode: both;
      animation-duration: #{getVar('colorWall.transitionTime')}ms;
      animation-timing-function: cubic-bezier(0.25, 0.46, 0.45, 0.94);
    }

    &-enter-done {
      backface-visibility: unset;
      animation: none;
      transform: unset;
    }

    &--exit,
    &--exit-active,
    &--exit-done {
      // zoomed color wall disappearing
      backface-visibility: hidden;
      position: absolute;
      top: 0;
      width: 100%;
      z-index: 1;
      will-change: transform;

      animation-name: zoomSwatchOut;
      animation-fill-mode: both;
      animation-duration: #{getVar('colorWall.transitionTime')}ms;
      animation-timing-function: cubic-bezier(0.25, 0.46, 0.45, 0.94);
    }

    &--exit-done {
      display: none;
    }
  }

  &__zoom-out {
    &--enter,
    &--enter-active,
    &--enter-done {
      // full wall appearing
      position: relative;
      z-index: 0;
    }

    &--exit,
    &--exit-active,
    &--exit-done {
      // full wall disappearing
      position: absolute;
      top: 0;
      width: 100%;
      z-index: 0;
    }

    &--exit-done {
      display: none;
    }
  }
}

// END COLOR WALL <-> COLOR SWATCH ACTIVE TRANSITION
// ------------------------------------------------------------------