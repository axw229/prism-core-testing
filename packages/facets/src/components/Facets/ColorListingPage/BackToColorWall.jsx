// @flow
import React, { useEffect, useState } from 'react'
import { FormattedMessage } from 'react-intl'
import { useSelector } from 'react-redux'
import { useParams } from 'react-router-dom'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { ROUTE_PARAMS } from 'constants/globals'
import { flattenDeep } from 'lodash'
import { type ColorsState } from 'src/shared/types/Actions.js.flow'
import 'src/providers/fontawesome/fontawesome'
import { fullColorName, generateColorWallPageUrl } from '../../../shared/helpers/ColorUtils'
import { urlWorker } from '../../../shared/helpers/URLUtils'
import { type Color } from '../../../shared/types/Colors'
import ButtonBar from '../../GeneralButtons/ButtonBar/ButtonBar'
import { MODE_CLASS_NAMES } from '../ColorWall/shared'

function findMatchIn(lookIn: Array<string | number> = [], lookFor: string | number) {
  if (lookIn.length === 0) return false
  if (typeof lookFor === 'undefined') return false

  const lookForAsString = String(lookFor)

  return lookIn.reduce((accum, next) => {
    if (accum) return true
    if (String(next) === lookForAsString) return true
  }, false)
}

function getWallColorIds(val: any): Array<number | string> {
  if (val.type === 'CHUNK') {
    return val.children
  }
  if (val.children) {
    return flattenDeep(val.children.map(getWallColorIds))
  }
  return []
}

export function BackToColorWall() {
  const {
    items: { colorMap = {} },
    status: { requestComplete, error },
    cwv3,
    layouts,
    groups,
    subgroups,
    shapes,
    family,
    section,
    search,
    structure
  }: ColorsState = useSelector((state) => state.colors)
  const { colorId }: { colorId?: ?string } = useParams()
  const color: Color = colorId && colorMap[colorId]

  const [backPath, setBackPath] = useState('')

  useEffect(() => {
    if (colorId && requestComplete && !error && color) {
      if (!cwv3) {
        // non-CWv3, legacy color wall
        // !cwv3 => colors > layouts { name: <section> } > families { name: <family> } > flattenDeep(unChunkedChunks)
        // !cwv3 => colors > structure { default: true }.name <-- use as link for "Back to wall"
        if (section) {
          const foundSection = layouts.reduce((accum, next) => {
            if (accum) return accum
            if (next.name === section) return next
          }, null)

          if (family && foundSection) {
            // find ID in family
            // if the above fails, and we detect this is a bright, look for its bright in this family
            const foundFamily = foundSection.families.reduce((accum, next) => {
              if (accum) return accum
              if (next.name === family) return next
            }, null)

            if (foundFamily) {
              // flatten its unChunkedChunks to match on colorId
              // if match, use foundfamily and section to form the back link
              if (findMatchIn(flattenDeep(foundFamily.unChunkedChunks), colorId)) {
                // we found one in family! do something about it
                setBackPath(
                  generateColorWallPageUrl(
                    foundSection.name,
                    foundFamily.name,
                    colorId,
                    fullColorName(color.brandKey, color.colorNumber, color.name)
                  )
                )
                return () => {}
              }

              if (findMatchIn(flattenDeep(foundFamily.unChunkedChunks), `bright-${colorId}`)) {
                // we found one in family! do something about it
                setBackPath(
                  generateColorWallPageUrl(
                    foundSection.name,
                    foundFamily.name,
                    `bright-${colorId}`,
                    fullColorName(color.brandKey, color.colorNumber, color.name)
                  )
                )
                return () => {}
              }
            }
          }

          if (foundSection) {
            // find ID in section
            if (findMatchIn(flattenDeep(foundSection.unChunkedChunks), colorId)) {
              // we found one in section! do something about it
              setBackPath(
                generateColorWallPageUrl(
                  foundSection.name,
                  null,
                  colorId,
                  fullColorName(color.brandKey, color.colorNumber, color.name)
                )
              )
              return () => {}
            }

            if (findMatchIn(flattenDeep(foundSection.unChunkedChunks), `bright-${colorId}`)) {
              // we found one in section! do something about it
              setBackPath(
                generateColorWallPageUrl(
                  foundSection.name,
                  null,
                  `bright-${colorId}`,
                  fullColorName(color.brandKey, color.colorNumber, color.name)
                )
              )
              return () => {}
            }
          }

          // if the above fails and we detect this is a bright, look for its bright in this family
        }

        const defaultSection =
          structure.reduce((accum, next) => {
            if (accum) return accum
            if (next.default) return next
            return accum
          }, null) || structure[0]

        if (defaultSection) {
          const foundSection = layouts.reduce((accum, next) => {
            if (accum) return accum
            if (next.name === defaultSection.name) return next
          }, null)

          if (foundSection) {
            if (findMatchIn(flattenDeep(foundSection.unChunkedChunks), colorId)) {
              setBackPath(
                generateColorWallPageUrl(
                  foundSection.name,
                  null,
                  colorId,
                  fullColorName(color.brandKey, color.colorNumber, color.name)
                )
              )
              return () => {}
            }

            if (findMatchIn(flattenDeep(foundSection.unChunkedChunks), `bright-${colorId}`)) {
              setBackPath(
                generateColorWallPageUrl(
                  foundSection.name,
                  null,
                  `bright-${colorId}`,
                  fullColorName(color.brandKey, color.colorNumber, color.name)
                )
              )
              return () => {}
            }
          }

          setBackPath(generateColorWallPageUrl(defaultSection.name))
          return () => {}
        }
      } else {
        // CWv3, modern color wall
        // cwv3 => colors > groups { name: <section> } > subgroups { name: <family> } > shapeId... but shape will need a lot of work to extract colors
        // cwv3 => colors > groups { default: true }.name <-- use as link for "back to wall"
        if (section) {
          const foundSection = groups.reduce((accum, next) => {
            if (accum) return accum
            if (next.name === section) return next
          }, null)

          if (family && foundSection) {
            // find ID in family
            // if the above fails, and we detect this is a bright, look for its bright in this family
            const subgroupMapping = foundSection.subgroups
              .map((sgId) => subgroups.filter((sg) => sg.id === sgId)[0])
              .filter(Boolean)

            const foundFamily = subgroupMapping.reduce((accum, next) => {
              if (accum) return accum
              if (next.name === family) return next
            }, null)

            if (foundFamily) {
              const famShape = shapes.filter((v) => v.id === foundFamily.shapeId)[0]
              const colorIds = famShape ? getWallColorIds(famShape.shape) : null
              // flatten its unChunkedChunks to match on colorId
              // if match, use foundfamily and section to form the back link
              if (colorIds && findMatchIn(colorIds, colorId)) {
                // we found one in family! do something about it
                setBackPath(
                  generateColorWallPageUrl(
                    foundSection.name,
                    foundFamily.name,
                    colorId,
                    fullColorName(color.brandKey, color.colorNumber, color.name)
                  )
                )
                return () => {}
              }

              if (colorIds && findMatchIn(colorIds, `bright-${colorId}`)) {
                // we found one in family! do something about it
                setBackPath(
                  generateColorWallPageUrl(
                    foundSection.name,
                    foundFamily.name,
                    `bright-${colorId}`,
                    fullColorName(color.brandKey, color.colorNumber, color.name)
                  )
                )
                return () => {}
              }
            }
          }

          if (foundSection) {
            const sectShape = shapes.filter((v) => v.id === foundSection.shapeId)[0]
            const colorIds = sectShape ? getWallColorIds(sectShape.shape) : null
            // find ID in section
            if (colorIds && findMatchIn(colorIds, colorId)) {
              // we found one in section! do something about it
              setBackPath(
                generateColorWallPageUrl(
                  foundSection.name,
                  null,
                  colorId,
                  fullColorName(color.brandKey, color.colorNumber, color.name)
                )
              )
              return () => {}
            }

            if (colorIds && findMatchIn(colorIds, `bright-${colorId}`)) {
              // we found one in section! do something about it
              setBackPath(
                generateColorWallPageUrl(
                  foundSection.name,
                  null,
                  `bright-${colorId}`,
                  fullColorName(color.brandKey, color.colorNumber, color.name)
                )
              )
              return () => {}
            }
          }

          // if the above fails and we detect this is a bright, look for its bright in this family
        }

        const defaultSection =
          groups.reduce((accum, next) => {
            if (accum) return accum
            if (next.default) return next
            return accum
          }, null) || groups[0]

        if (defaultSection) {
          const sectShape = shapes.filter((v) => v.id === defaultSection.shapeId)[0]
          const colorIds = sectShape ? getWallColorIds(sectShape.shape) : null

          if (colorIds && findMatchIn(colorIds, colorId)) {
            setBackPath(
              generateColorWallPageUrl(
                defaultSection.name,
                null,
                colorId,
                fullColorName(color.brandKey, color.colorNumber, color.name)
              )
            )
            return () => {}
          }

          if (colorIds && findMatchIn(colorIds, `bright-${colorId}`)) {
            setBackPath(
              generateColorWallPageUrl(
                defaultSection.name,
                null,
                `bright-${colorId}`,
                fullColorName(color.brandKey, color.colorNumber, color.name)
              )
            )
            return () => {}
          }

          setBackPath(generateColorWallPageUrl(defaultSection.name))
          return () => {}
        }
      }
    }
  }, [colorId, color, section, family, cwv3, layouts, groups, subgroups, shapes, structure, requestComplete, error])

  const url = search.active && search.query ? urlWorker.set(ROUTE_PARAMS.SEARCH, search.query).in(backPath) : backPath
  return (
    <div className={MODE_CLASS_NAMES.BASE}>
      <div className={MODE_CLASS_NAMES.COL}>
        <div className={MODE_CLASS_NAMES.CELL}>
          <ButtonBar.Bar>
            <ButtonBar.Button to={url}>
              <FontAwesomeIcon icon={['fal', 'th-large']} pull='left' fixedWidth />
              <FormattedMessage id='BACK_TO_COLOR_WALL' />
            </ButtonBar.Button>
          </ButtonBar.Bar>
        </div>
      </div>
    </div>
  )
}

export default BackToColorWall
