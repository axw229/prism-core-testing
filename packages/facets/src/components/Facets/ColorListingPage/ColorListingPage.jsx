// @flow
import React, { useContext } from 'react'
import facetBinder from 'src/facetSupport/facetBinder'
import { Switch, Route, useParams, useLocation } from 'react-router-dom'
import { ColorWallPage } from '../ColorWallFacet'
import BackToColorWall from './BackToColorWall'
import { ColorDetailsPage } from 'src/components/Facets/ColorDetailsFacet'
import { CSSTransition, SwitchTransition } from 'react-transition-group'
import './ColorListingPage.scss'
import ConfigurationContext from '../../../contexts/ConfigurationContext/ConfigurationContext'
import { useIntl } from 'react-intl'

const ColorWallCDP = () => {
  const { locale } = useIntl()
  const { brandId } = useContext(ConfigurationContext)
  return (
    <div className='inline-color-details'>
      <BackToColorWall />
      <ColorDetailsPage colorSEO={useParams().colorSEO} brand={brandId} language={locale} />
    </div>
  )
}

type ColorListingPageProps = {
  brand: string,
  language: string
}

export default facetBinder((props: ColorListingPageProps) => {
  const location = useLocation()
  return (
    <div className='transition-wrapper'>
      <SwitchTransition mode='out-in'>
        <CSSTransition key={location.pathname.indexOf('color-detail')} addEndListener={(node, done) => node.addEventListener('transitionend', done, false)}>
          <Switch location={location}>
            <Route path='/active/color-detail/:colorId/:colorSEO' render={() => <ColorWallCDP language={props.language} brand={props.brand} />} />
            <Route render={() => <ColorWallPage resetOnUnmount={false} displayDetailsLink displayAddButton={false} />} />
          </Switch>
        </CSSTransition>
      </SwitchTransition>
    </div>
  )
}, 'ColorListingPage')
