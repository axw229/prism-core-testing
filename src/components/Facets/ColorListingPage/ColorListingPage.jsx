// @flow
import React, { PureComponent } from 'react'
import { Route, Redirect, withRouter, Switch } from 'react-router-dom'
import { TransitionGroup, CSSTransition } from 'react-transition-group'
// TODO: PRISM-370 | Facets are top-level components -- refactor ColorWallPage so it can be imported without facet-related functionality @cody.richmond
import { ColorWallPage } from '../ColorWallFacet'
import ColorDetails from '../ColorDetails/ColorDetails'
import BackToColorWall from './BackToColorWall'
import facetBinder from 'src/facetSupport/facetBinder'
import includes from 'lodash/includes'
import { varValues } from 'src/shared/variableDefs'
import { ROUTE_PARAMS, ROUTE_PARAM_NAMES } from 'constants/globals'
import { facetBinderDefaultProps, type FacetBinderMethods } from 'src/facetSupport/facetInstance'
import { type FacetPubSubMethods, facetPubSubDefaultProps } from 'src/facetSupport/facetPubSub'
import './ColorListingPage.scss'

const colorWallBaseUrl = `/${ROUTE_PARAMS.ACTIVE}/${ROUTE_PARAMS.COLOR_WALL}`
const colorDetailsBaseUrl = `/${ROUTE_PARAMS.ACTIVE}/${ROUTE_PARAMS.COLOR_DETAIL}`

// this is very vague because react-router doesn't have the ability to match /section/x/family/y/color/z and /section/x/color/z with the same route
// we're handling the URL-parsing logic manually in ColorWallComponent below
const colorWallUrlPattern = `${colorWallBaseUrl}(/.*)?`

// barebones component to always take the user to active if they try to access root.
// not sure if we need this but if we end up using this for TAG & want to retain bookmarks..
export const RootRedirect = () => {
  return <Redirect to={colorWallBaseUrl} />
}

// since the CDP component won't have any color information if we go to it directly, we need to wrap it
// in the ColorDataWrapper HOC to ensure it has color data prior to rendering it.
export const ColorDetailsComponent = (props: any) => {
  // need a wrapping element that isn't a Fragment in order to get the transition classes applied to the group
  return (
    <div>
      <BackToColorWall />
      <ColorDetails {...props} />
    </div>
  )
}

const ColorWallPageNoReset = () => <ColorWallPage resetOnUnmount={false} />

type ColorListingPageProps = FacetBinderMethods & FacetPubSubMethods & { location: Object }

type ColorListingPageState = { prevPathname?: string, toDetails: boolean, toWall: boolean }

export class ColorListingPage extends PureComponent<ColorListingPageProps, ColorListingPageState> {
  static defaultProps = {
    ...facetBinderDefaultProps,
    ...facetPubSubDefaultProps
  }
  state: ColorListingPageState = { prevPathname: void (0), toDetails: false, toWall: false }

  render () {
    const { location } = this.props
    const { toWall, toDetails } = this.state
    let transitionClassNames = 'cdp-slide'

    if (toWall) {
      transitionClassNames = 'cdp-slide-to-wall cdp-slide'
    } else if (toDetails) {
      transitionClassNames = 'cdp-slide-to-details cdp-slide'
    }

    // grabbing the first half of the path to use as a key so we only actually trigger
    // a CSS transition when /active/color-wall/ changes to /active/color/ and vice-versa
    const key = location.pathname.split('/').slice(2, 3).join()

    return (
      <React.Fragment>
        {/* Keeping this Route outside of TransitionGroup in order to avoid introducing MaximumUpdateDepth errors upon initial redirect */}
        <Switch location={location}>
          <Route path='/' exact component={RootRedirect} />
        </Switch>
        <TransitionGroup className='color-wall-transitioner'>
          <CSSTransition
            key={key}
            classNames={transitionClassNames}
            timeout={varValues.colorWall.transitionTime * 1.2}
          >
            <Switch location={location}>
              <Route path={colorWallUrlPattern} component={ColorWallPageNoReset} />
              <Route path={`${colorDetailsBaseUrl}/:${ROUTE_PARAM_NAMES.COLOR_ID}/:${ROUTE_PARAM_NAMES.COLOR_SEO}`} exact component={ColorDetailsComponent} />
            </Switch>
          </CSSTransition>
        </TransitionGroup>
      </React.Fragment>
    )
  }

  static getDerivedStateFromProps (props: ColorListingPageProps, state: ColorListingPageState) {
    const { prevPathname } = state
    const { location: { pathname: newPropsPathname } } = props

    if (newPropsPathname !== prevPathname) {
      let newProps = { prevPathname: newPropsPathname }

      if (includes(newPropsPathname, colorDetailsBaseUrl) && includes(prevPathname, colorWallBaseUrl)) {
        // going TO color details FROM color wall; slide in from right
        newProps = Object.assign({}, newProps, { toDetails: true, toWall: false })
        // transitionClassNames = ''
      } else if (includes(newPropsPathname, colorWallBaseUrl) && includes(prevPathname, colorDetailsBaseUrl)) {
        // going TO color wall FROM color details; slide in from left
        newProps = Object.assign({}, newProps, { toDetails: false, toWall: true })
      }
      return newProps
    }
    return null
  }
}

export default facetBinder(withRouter(ColorListingPage), 'ColorListingPage')
