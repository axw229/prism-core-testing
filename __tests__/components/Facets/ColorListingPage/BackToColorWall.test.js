import React from 'react'
import { BackToColorWall } from 'src/components/Facets/ColorListingPage/BackToColorWall'
import * as Colors from '__mocks__/data/color/Colors'
import { FormattedMessage } from 'react-intl'
import { Link } from 'react-router-dom'

describe('BackToColorWall', () => {
  let backToColorWall

  beforeAll(() => {
    // eslint-disable-next-line no-undef
    backToColorWall = mocked(<BackToColorWall color={Colors.getColor()} />)
  })

  it('should rendring format message component', () => expect(backToColorWall.find(FormattedMessage)).toBeTruthy())

  it('should rendring Link correctly', () => expect(backToColorWall.find(Link)).toBeTruthy())
})
