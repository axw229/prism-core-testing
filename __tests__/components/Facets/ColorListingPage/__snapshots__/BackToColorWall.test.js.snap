// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot match testing should match snapshot 1`] = `
<div
  className="color-wall-mode-btns"
>
  <div
    className="color-wall-mode-btns__col"
  >
    <div
      className="color-wall-mode-btns__cell"
    >
      <Bar>
        <Button
          to="/active/color-wall/color/2248/sw-6561-teaberry/"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={true}
            flip={null}
            icon={
              Array [
                "fal",
                "th-large",
              ]
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull="left"
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
          <FormattedMessage
            id="BACK_TO_COLOR_WALL"
            values={Object {}}
          />
        </Button>
      </Bar>
    </div>
  </div>
</div>
`;
